compiler/rustc_trait_selection/src/solve/normalizes_to/mod.rs

```diff
@@ -379,6 +379,45 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
     ) -> QueryResult<'tcx> {
         let tcx = ecx.tcx();
         ecx.probe_misc_candidate("builtin pointee").enter(|ecx| {
+            fn handle_product_ty<'tcx>(
+                ecx: &mut EvalCtxt<'_, 'tcx>,
+                goal: Goal<'tcx, NormalizesTo<'tcx>>,
+                tail_ty: Ty<'tcx>,
+            ) -> QueryResult<'tcx> {
+                let tcx = ecx.tcx();
+
+                // If `Self: Sized`, then `<Self as Pointee>::Metadata == ()`.
+                let sized_probe =
+                    ecx.probe_misc_candidate("builtin pointee via sized").enter(|ecx| {
+                        let sized_predicate = ty::TraitRef::from_lang_item(
+                            tcx,
+                            LangItem::Sized,
+                            DUMMY_SP,
+                            [ty::GenericArg::from(goal.predicate.self_ty())],
+                        );
+                        // FIXME(-Znext-solver=coinductive): Should this be `GoalSource::ImplWhereBound`?
+                        ecx.add_goal(GoalSource::Misc, goal.with(tcx, sized_predicate));
+                        ecx.eq(goal.param_env, goal.predicate.term, tcx.types.unit.into())
+                            .expect("expected goal term to be fully unconstrained");
+                        ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                    });
+
+                if let Ok(response) = sized_probe {
+                    return Ok(response);
+                }
```

Comment:
actually, this isn't as trivial as I originally thought. Builtin traits should generally follow the same rules as non-builtin traits. If they don't, it's very easy to accidentally break an invariant of the type system.

Right now you pretty much add these builtin impls
```rust
struct MyType<T: ?Sized> {
    last: Tail<T>,
}

impl<T: ?Sized> Pointee for MyType<T>
where
    Self: Sized,
{
    type Metadata = ();
}

impl<T: ?Sized> Pointee for MyType<T> {
    type Metadata = <Tail<T> as Pointee>::Assoc;
}
```
These impls overlap. While I can't think of concrete problem caused by this overlap, it should still be avoided if possible. We do not allow overlapping impls with associated types anywhere else (apart from the "dyn Trait: Trait" impl, which is currently causing an actual soundness bug #57893). Now, the associated type of these two impls should always end up being the same, so it probably does not cause an unsoundness that way. But having impls like this can also result in ambiguity. Regardless, it complicates the reasoning about the type system.

Thinking about this, we should be able to only emit the second impl, i.e.
```rust
impl<T: ?Sized> Pointee for Adt {
    type Metadata = <Tail<T> as Pointee>::Assoc;
}
```
Because types cannot be recursive without any indirection, this should bottom out in a type for which we can trivially compute the metadata.

Reply:
> Is this expected behavior? With manual impls you get an ambiguity error:

no, this is because of your implementation. This way of handling multiple candidates (by only considering the second one if the first one does not apply) is something we avoid elsewhere. We instead collect all candidates and then use `try_merge_candidates` to actually get to a single one.

> You could argue that's fine to break, because `Pointee` is unstable and nobody writes code like that anyway. However, breaking this code would create situations where it's possible to cast a thin pointer to `*const Wrapper<Tail>`, but not know `Wrapper<Tail>: Thin`. The entire purpose of this PR is bringing the checks of pointer-to-pointer casts and the builtin impl of the `Pointee` trait closer together

I agree that it is weird to be able to pointer cast while not actually being able to prove that the cast is correct in the type system. However, I would still really like to avoid these overlapping impls. Can pointer casting be changed to rely on `Metadata` being equal instead of using a separately written check?

Reply:
> no, this is because of your implementation.

This compiles on nightly, without my implementation.

```rust
#![feature(ptr_metadata)]
use std::ptr::Pointee;

fn check_meta<T, U: Pointee<Metadata = T>>() {}

fn sized_slice<T>()
where
    [T]: Sized,
{
    // We have `[T]: Sized`, but the metadata is still `usize`.
    check_meta::<usize, [T]>(); // compiles
}
```

What I'm trying to say is that if my implementation is wrong, we should change this to be an ambiguity error as well.

---

> Can pointer casting be changed to rely on `Metadata` being equal instead of using a separately written check?

Checking for equal metadata alone is not enough, because that can't cover fat->thin casts or casts between pointers to trait objects that change the trait object lifetime or add auto traits. We'd need something like a new built-in trait `MetadataCompatible<T>` (or new predicate kind) to fully typecheck pointer-to-pointer casts with traits. I've actually been working on a [proposal](https://hackmd.io/BcsyukxVSm-mLPY4Rlj5mg) and [experimental implementation](https://github.com/lukas-code/rust/pull/6) to move pointer-to-pointer casts to the trait system entirely. This PR is the first step to making that happen and I was hoping it would be sufficiently motivated by just the `cast_same_meta` example from the OP and "let's not break code that currently compiles", but evidently that's not the case, so here we are.

---

Would it perhaps be possible to treat the current implementation as a specialization, i.e.

```rust
impl<T: ?Sized> Pointee for MyType<T>
where
    T: Pointee,
{
    default type Metadata = <T as Pointee>::Metadata;
}

impl<T: ?Sized> Pointee for MyType<T>
where
    T: Pointee,
    T: Sized,
{
    type Metadata = ();
}
```

That appears to compile on nightly, but idk if it's actually sound.

Otherwise, I guess just normalizing to the metadata of the tail is an improvement over the status quo and all cases that it can break are easily fixed by adding `where Tail: Sized`.

Reply:
(clarifying previous comments)

The current impls for `Pointee` on nightly looks something like this:

```rust
impl Pointee for u8 {
    type Metadata = ();
}

impl Pointee for [u8] {
    type Metadata = usize;
}

impl Pointee for Wrapper<u8> {
    type Metadata = ();
}

impl Pointee for Wrapper<[u8]> {
    type Metadata = usize;
}

// this impl is only selected if no other candidate applies
fallback impl<T: Sized> Pointee for T {
    type Metadata = ();
}
```

That means that if we know `T == [u8]` and `T: Sized`, then `T::Metadata == usize`, but if we only know `T: Sized`, then `T::Metadata == ()`.

The impls in this PR as is will extend the above by another fallback case:
```rust
impl Pointee for u8 {
    type Metadata = ();
}

impl Pointee for [u8] {
    type Metadata = usize;
}

impl Pointee for Wrapper<u8> {
    type Metadata = ();
}

impl Pointee for Wrapper<[u8]> {
    type Metadata = usize;
}

// this impl is only selected if no candidate above applies
fallback impl<T: Sized> Pointee for T {
    type Metadata = ();
}

// this impl is only selected if no candidate above applies, including the other fallback
fallback fallback impl<T: ?Sized> Pointee for Wrapper<T> {
    type Metadata = <T as Pointee>::Metadata;
}
```

An alternative approach that doesn't rely on such fallbacks, which can't currently be expressed in the type system, would instead rely on specialization:

```rust
impl Pointee for u8 {
    default type Metadata = ();
}

impl Pointee for [u8] {
    default type Metadata = usize;
}

impl<T: ?Sized> Pointee for Wrapper<T> {
    default type Metadata = <T as Pointee>::Metadata;
}

// this impl is preferred over all other impls
specialize impl<T: Sized> Pointee for T {
    type Metadata = ();
}
```

With that, if we know `T == [u8]` and `T: Sized`, then `T::Metadata == ()`, because the specialization will take priority over the defaults.